### ISSUE TITLE
When rubyzip is used to modify a DOCX file, the resulting file can't be opened

### ISSUE THREAD

**Issue by alexdowad** [rubyzip/rubyzip#15]
Take a DOCX file (which is just a ZIP file with a different extension). Use Zip::ZipFile to remove one file from the archive, and replace it with an exact, identical one. The resulting file won't be openable by Microsoft Word, OpenOffice, etc.

require "zip/zipfilesystem"
f1 = File.binread("test.docx")
Zip::ZipFile.open("test.docx") { |z| s = z.file.read("word/document.xml"); z.file.delete("word/document.xml"); z.open("word/document.xml","w") { |f| f.write(s) } }
f2 = File.binread("test.docx")
f1 == f2   => false
f1.bytes.take(15)
=> [80, 75, 3, 4, 20, 0, 6, 0, 8, 0, 0, 0, 33, 0, 111]
f2.bytes.take(15)
=> [80, 75, 3, 4, 10, 0, 0, 0, 8, 0, 0, 0, 33, 0, 111]

After the ZIP file is read and written by rubyzip, a couple of bytes in the header are changed. I don't know enough about the ZIP format to understand why. But when I do the same thing with a standalone ZIP utility, it doesn't do that, and the resulting DOCX file can be opened properly.

**Comment by tinabel** [comment]
I'm still having a problem getting docx files to open correctly after zipping them with RubyZip. Is this issue going to be fixed or are we left with jondruse's code as the only solution to the issue?

**Comment by tinabel** [comment]
I was able to solve this using [this code](https://gist.github.com/tinabel/ddd5cc9b0dd762986918520a132800d2), though I ultimately took the approach of shelling zip out to the command line, as it was a more direct solution for my needs.

**Comment by hainesr** [comment]
@tinabel I have been unable to replicate the issue with the code you posted above - i.e. if I remove the `docx` specific bit in `put_into_archive` and use your code to recursively zip a directory, I can unzip the generated archive and there are no differences between the original files and their zipped/unzipped counterparts. I assume I'm missing something; do you have an example `docx` file that triggers the issue?

Also which platform and version of rubyzip are you using?

Thanks.

**Comment by tanin47** [comment]
> I was able to get docx files to work correctly by using write_buffer() instead of open(). Maybe that will help lead to the issue. But my working example is here;
> 
> https://gist.github.com/e7d2855435654e1ebc52

I want to help confirm that this works for me. I simply reopen the corrupted zip and rewrite it again using the above code, and it works.

Here's my code:

```
Zip::File.open(output_file_path) do |zip_file|
  zip_file.get_output_stream(zip_file.find_entry('word/document.xml')) { |f| f.write(@doc_root.to_xml) }
  # write more stuff
end

# At this point, the docx is still corrupted

buffer = Zip::File.open(output_file_path) do |original|
  Zip::OutputStream.write_buffer do |out|
    original.entries.each do |e|
      if e.ftype == :directory
        out.put_next_entry(e.name)
      else
        out.put_next_entry(e.name)
        out.write e.get_input_stream.read
      end
    end
  end
end

File.open(output_file_path, "wb") {|f| f.write(buffer.string) }

# Now the docx is valid.
```

This only happens to some docx files. I've encountered a docx file where I didn't need to do this.

**Comment by hainesr** [comment]
Thanks for this @tanin47. As you can see from the above thread, I have struggled to replicate these behaviours.

Please could you let me know which platform you are on, and which version of rubyzip you are using? And if you have an example docx file that triggers these behaviours, that would be even better thanks!

**Comment by tanin47** [comment]
@hainesr MacOS with Microsoft Word. I have one docx that doesn't encounter this bug and another docx that does. This would be a good example! I'll need to anonymize the doc first.

I don't know what the difference is between 2 docs. I didn't look at it too deeply.

Let me get back to you soon.

**Comment by tanin47** [comment]
If I open the file with Microsoft Word, modify to replace all PIIs, and save, then the file isn't problematic anymore. Give me a bit more time.

**Comment by hainesr** [comment]
Thanks for checking!

Are you using the latest release of RubyZip (3.2.2)?

**Comment by tanin47** [comment]
Yes (based on my gemlock).

I'm unable to produce a corrupted doc. I tried creating a doc on windows and sending it to my mac and use rubyzip. It didn't encounter this bug.

2 of my separate customers sent docs that were corrupted. But if I open them and add one character and save, the doc becomes unproblematic and doesn't encounter the issue I described. Unfortunately, I can't share the docs from my customer.

In the future, if I'm able to obtain a document that can be shared, I'll come back!
